### PR TITLE
Added more hooks and callbacks and implemented global learning rate scheduler.

### PR DESCRIPTION
**Pull request by cying17**
This is a follow-up PR of PR #222 to add hooks and callbacks for the basic server and client.

It also implemented the global learning rate scheduler. 

## Description

As a starting point, the following hooks/callbacks have been implemented in this PR: 
For server: weights_aggregated, server_will_close
For client: client_train_end

More hooks/callbacks will be implemented in the future as the need arises.

Also, to showcase using the new hooks and callbacks, examples/oort has been revised and examples/fedscr has been implemented to avoid cutting and pasting most of the code in the training loop.

If in the configuration file, `trainer.global_lr_scheduler` is `true`, the learning rate of the first epoch in the next communication round is scheduled based on that of the last epoch in the previous communication round.

## How has this been tested?

`./run -c configs/MNIST/fedavg_lenet5.yml`

`python examples/oort/oort.py -c examples/oort/oort_MNIST_lenet5.yml`

`python examples/fedscr/fedscr.py -c examples/fedscr/fedscr_MNIST_lenet5.yml`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by baochunli** [comment]
The design of where to add new hooks and callbacks in the clients and server needs some discussions.

As a starting point, I don't think we need to add `on_server_selection_start` and other similar callbacks that are not used by Oort or other existing examples. We do need to add a callback in `aggregate_weights` since both Oort and Port were copying and pasting code from this function. Perhaps this callback can be called `on_weights_aggregated`.

Also, it would be a good practice to add a corresponding hook for each callback. For example, we may need a hook called `weights_aggregated`.

In Oort, we can change the name of the private functions (such as `update_clients`) just to make sure that these are recognized as private functions within the class, rather than inherited from the parent class.

File: examples/fedscr/fedscr_server.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            model_name = Config().trainer.model_name
+
+            if not os.path.exists(checkpoint_path):
+                os.makedirs(checkpoint_path)
+
+            loss_path = f"{checkpoint_path}/{model_name}_thresholds.pkl"
+
+            with open(loss_path, "wb") as file:
+                pickle.dump(self.update_threshold, file)
+
```
Comment (baochunli) [inline examples/fedscr/fedscr_server.py:168]: The way that FedSCR implements passing information (thresholds in this case) from server to the clients is completely wrong. It used files, which must be changed to using the communication channel.

**Review by baochunli — CHANGES_REQUESTED** [review]

File: examples/fedscr/fedscr_server.py (hunk 1 above)
Comment (cying17) [reply examples/fedscr/fedscr_server.py]: Changed in commit 63b2bb9ed59f74680d1d7028943363636f48c0d0.
